### PR TITLE
A12: Wordle

### PR DESCRIPTION
**Pull request by Dakshita-Minocha**

### CODE REVIEW

File: Training/Program.cs [hunk 1]
```diff
@@ -3,31 +3,202 @@
...
+   int mGuessNum;
+   #endregion
+
+   #region Implementation ----------------------------------------
+   /// <summary>Checks if input is a valid word from dictionary,
+   /// compares input with generated Secret Word, and re-prints coloured input as hints.</summary>
+   /// <returns> Null if input is not a valid word.
+   /// False if input is a valid wrd but is not the secret word.
+   /// True if input is the secret word.</returns>
+   bool? CompareAndPrint (char[] input) {
```
Comment (asfar-nazar) [inline Training/Program.cs:72]: input can be a string itself as it is an IEnumerable<char>

File: Training/Program.cs [hunk 2]
```diff
@@ -3,31 +3,202 @@
...
+#region Class Wordle ---------------------------------------------------------------------------
+public class Wordle {
+   #region Constructors ------------------------------------------
+   /// <summary>Static constructor to read and close dict files.</summary>
+   static Wordle () {
+      mPuzzleWords = File.ReadAllLines ("C:/etc/Puzzle-5.txt");
+      mDict = File.ReadAllLines ("C:/etc/Dict-5.txt");
+      CursorVisible = false;
+   }
+   static string[] mPuzzleWords, mDict;
```
Comment (asfar-nazar) [inline Training/Program.cs:33]:
```suggestion
   static string[] mPuzzleWords, mDictionaryWords;
```
For clarity

File: Training/Program.cs [hunk 3]
```diff
@@ -3,31 +3,202 @@
...
+   #endregion
+
+   #region Implementation ----------------------------------------
+   /// <summary>Checks if input is a valid word from dictionary,
+   /// compares input with generated Secret Word, and re-prints coloured input as hints.</summary>
+   /// <returns> Null if input is not a valid word.
+   /// False if input is a valid wrd but is not the secret word.
+   /// True if input is the secret word.</returns>
+   bool? CompareAndPrint (char[] input) {
+      if (!mDict.Any (x => x.SequenceEqual (string.Join ("", input)))) {
```
Comment (asfar-nazar) [inline Training/Program.cs:73]:
```suggestion
      if (!mDict.Contains (string.Join ("", input))) {
```

File: Training/Program.cs [hunk 4]
```diff
@@ -3,31 +3,202 @@
...
+         mState[input[i]] = ForegroundColor = input[i] == mSecretWord[i] ? ConsoleColor.Green :
+                           (mSecretWord.Any (x => x == input[i]) ? ConsoleColor.Blue : ConsoleColor.DarkGray);
+         Write ($"{input[i]}   ");
+      }
+      (int c, int r) = (mCurrentCol, mRow);
+      (mCurrentCol, mRow) = (mStartCol - mMaxWordLen, mAlphaRow);
+      ShiftTo (mCurrentCol, mAlphaRow);
+      PrintAlphabets ();
+      (mCurrentCol, mRow) = (c, r);
+      if (input.SequenceEqual (mSecretWord)) return true;
```
Comment (asfar-nazar) [inline Training/Program.cs:88]:
```suggestion
      return input == mSecretWord;
```

File: Training/Program.cs [hunk 5]
```diff
@@ -3,31 +3,202 @@
...
+      return false;
```
Comment (asfar-nazar) [inline Training/Program.cs:89]:
```suggestion

```

File: Training/Program.cs [hunk 6]
```diff
@@ -3,31 +3,202 @@
...
+      NextLine ();
+      ResetColor ();
+      Message (mCorrectGuess == true ?
+               $"Congrats! You got it right in just {6 - mGuessNum} guess{(mGuessNum == 5 ? "" : "es")}." :
+               $"Oops! The answer was {string.Join ("", mSecretWord)}. Better luck next time!");
+      mState.Clear ();
+   }
+   static Random mRandomWord = new ();
+   char[] mSecretWord;
+   bool? mCorrectGuess;
```
Comment (asfar-nazar) [inline Training/Program.cs:62]: We can use a normal bool here. In GetInput method, we can remove the check for null as mCursor <= 5 performs the same check ie user is still typing a guess and has not hit enter

File: Training/Program.cs [hunk 7]
```diff
@@ -3,31 +3,202 @@
...
+         return null;
       }
-      #endregion
+      ShiftTo (mStartCol, mRow);
+      for (int i = 0; i < mCursor; i++) {
+         mState[input[i]] = ForegroundColor = input[i] == mSecretWord[i] ? ConsoleColor.Green :
+                           (mSecretWord.Any (x => x == input[i]) ? ConsoleColor.Blue : ConsoleColor.DarkGray);
+         Write ($"{input[i]}   ");
+      }
+      (int c, int r) = (mCurrentCol, mRow);
```
Comment (asfar-nazar) [inline Training/Program.cs:83]: Is the swap to temp variables necessary here ? We can just call ShiftTo (mStartCol - mMaxWordLen, mAlphaRow)

File: Training/Program.cs [hunk 8]
```diff
@@ -3,31 +3,202 @@
...
+   /// True if input is the secret word.</returns>
+   bool? CompareAndPrint (char[] input) {
+      if (!mDict.Any (x => x.SequenceEqual (string.Join ("", input)))) {
+         Message ("Enter valid word.");
+         return null;
       }
-      #endregion
+      ShiftTo (mStartCol, mRow);
+      for (int i = 0; i < mCursor; i++) {
+         mState[input[i]] = ForegroundColor = input[i] == mSecretWord[i] ? ConsoleColor.Green :
```
Comment (asfar-nazar) [inline Training/Program.cs:79]:
```suggestion
         if (input[i] == mSecretWord[i]) mState[input[i]] = ForegroundColor = ConsoleColor.Green;
         else if (mSecretWord.Contains (input[i])) {
            if (input[..i].Count (input[i]) >= mSecretWord.Count (input[i])) mState[input[i]] = ForegroundColor = ConsoleColor.DarkGray;
            else mState[input[i]] = ForegroundColor = ConsoleColor.Blue;
         } else mState[input[i]] = ForegroundColor = ConsoleColor.DarkGray;
```
To print only the first 'p' in blue if secret word is 'tempt' and input is 'apple'

**Review by asfar-nazar — CHANGES_REQUESTED** [review]

File: Training/Program.cs (hunk 3 above)
Comment (Dakshita-Minocha) [reply Training/Program.cs:73]: I thought I'd use any as it'd exit as soon as it got something that matched?

Comment (asfar-nazar) [reply Training/Program.cs:73]: > I thought I'd use any as it'd exit as soon as it got something that matched?

Both 'Any' and 'Contains' will exit as soon as any of the elements match

File: Training/Program.cs (hunk 7 above)
Comment (Dakshita-Minocha) [reply Training/Program.cs:83]: Need to update value of currentCol and Row before and after use of PrintAlphabets(), these values are being used in NextLine() so it messes up the printing

File: TestTraining/TestTraining.cs [hunk 9]
```diff
@@ -5,13 +5,40 @@
...
+            w.UpdateState (ConsoleKey.Enter);
+            w.Save ();
+            w.mCursor = 0;
+            count += 26;
+         }
+         w.Save (-1);
+
+         string[] testLines = File.ReadAllLines (saveToFile[FilePtr++]);
+         Assert.IsTrue (testLines.SequenceEqual (refLines));
+         Assert.AreEqual (win, testLines[1]);
```
Comment (asfar-nazar) [inline TestTraining/TestTraining.cs:41]: This is not needed as the entire file is compared in the previous line

**Review by asfar-nazar — CHANGES_REQUESTED** [review]
See comment

**Review by asfar-nazar — APPROVED** [review]

**Pull request closed without merging**